gui/biomes.lua

```diff
@@ -0,0 +1,376 @@
+    RIVER_TROPICAL_BRACKISHWATER = "Tropical Brackishwater River",
+    RIVER_TROPICAL_SALTWATER = "Tropical Saltwater River",
+    SUBTERRANEAN_WATER = "Subterranean Water",
+    SUBTERRANEAN_CHASM = "Subterranean Chasm",
+    SUBTERRANEAN_LAVA = "Subterranean Lava",
+}
+
+local function find(t, predicate)
+    for k, item in pairs(t) do
+        if predicate(k, item) then
+            return k, item
+        end
+    end
+    return nil
+end
+
+local regionBiomeMap = {}
+local biomesMap = {}
+local biomeList = {}
+local function gatherBiomeInfo(z)
+    local maxX, maxY, maxZ = dfhack.maps.getTileSize()
+    maxX = maxX - 1; maxY = maxY - 1; maxZ = maxZ - 1
+
+    local z = z or df.global.window_z
+
+    --for z = 0, maxZ do
+    for y = 0, maxY do
+        for x = 0, maxX do
+            local rgnX, rgnY = dfhack.maps.getTileBiomeRgn(x,y,z)
+            if rgnX == nil then goto continue end
+
+            local regionBiomesX = regionBiomeMap[rgnX]
+            if not regionBiomesX then
+                regionBiomesX = {}
+                regionBiomeMap[rgnX] = regionBiomesX
+            end
+            local regionBiomesXY = regionBiomesX[rgnY]
+            if not regionBiomesXY then
+                regionBiomesXY = {
+                    biomeTypeId = dfhack.maps.getBiomeType(rgnX, rgnY),
+                    biome = dfhack.maps.getRegionBiome(rgnX, rgnY),
+                }
+                regionBiomesX[rgnY] = regionBiomesXY
+            end
+
+            local biomeTypeId = regionBiomesXY.biomeTypeId
+            local biome = regionBiomesXY.biome
+
+            local biomesZ = biomesMap[z]
+            if not biomesZ then
+                biomesZ = {}
+                biomesMap[z] = biomesZ
+            end
+            local biomesZY = biomesZ[y]
+            if not biomesZY then
+                biomesZY = {}
+                biomesZ[y] = biomesZY
+            end
+
+            local function currentBiome(_, item)
+                return item.biome == biome
+            end
+            local ix = find(biomeList, currentBiome)
+            if not ix then
+                local ch = string.char(string.byte('a') + #biomeList)
+                table.insert(biomeList, {biome = biome, char = ch, typeId = biomeTypeId})
+                ix = #biomeList
+            end
+
+            biomesZY[x] = ix
+
+            ::continue::
+        end
+    end
+    --end
+end
+
+-- always gather info at the very bottom first: this ensures the important biomes are
+-- always in the same order (high up in the air strange things happen)
+gatherBiomeInfo(0)
+
+--------------------------------------------------------------------------------
+
+local TITLE = "Biomes"
+
+if RELOAD then BiomeVisualizerLegend = nil end
+local DEFAULT_LIST_HEIGHT = 5
+BiomeVisualizerLegend = defclass(BiomeVisualizerLegend, widgets.Window)
+BiomeVisualizerLegend.ATTRS {
+    frame_title=TITLE,
+    frame_inset=0,
+    resizable=false,
```

Comment:
this can be left as true -- comments below

Reply:
made it resizable again with your suggestions